course5/homework03/homework03.cc

```diff
@@ -1,18 +1,68 @@
 #include <cstdio>
+#include <string>
 
-// TODO your code goes here
+/***
+Imagine that you are a young C++ developer, and your manager, for some reason, forbids std::cout. The only I/O function you are allowed to use to produce the output, is std::printf from <cstdio>
+
+You are not happy about it; you are using C++ after all, and you want to take advantage of the type-safety and convenience it offers.
+
+1. Write a library that provides you with “print”, a function you can use on a selection of types: int, bool, string, float, double etc. The function would use std::printf under the covers to print the value.
+2. For floating-point values, the function could take a second, optional parameter, integer, that specifies the precision with which the number is printed.
+***/
+namespace printer
+{
+	/***
+	provides overloaded function to print various types to stdout
+	***/
+	void print(int input)
+	{
+		std::printf("%i", input);
+	}
+
+	void print(double input, int precision = 0)
+	{
+		std::string format_str = "%." + std::to_string(precision) + "f";
+		const char * format_ptr = format_str.c_str();
+		std::printf(format_ptr, input);
+	}
+
+	void print(bool input)
+	{
+		if (input == 1)
+			std::printf("true");
+		else
+			std::printf("false");
+	}
+
+	void print(const char * input)
```

Comment:
At first I just had the signatures with parameters of type `const char` and `bool`, and it took me forever to get this right (c-style strings kept resolving to the bool-version). What I learned from StackOverflow: "'Hello World' is a string literal of type "array of 12 const char" which can be converted to a "pointer to const char" which can in turn be converted to a bool."

Thanks, compiler.